internal/authorizer/authorizer.go

```diff
@@ -0,0 +1,190 @@
+	for _, f := range args {
+		f(&opts)
+	}
+
+	return &Authorizer{
+		authority:   opts.authority,
+		credsLookup: opts.credLookup,
+		root:        opts.root,
+	}
+}
+
+type polkitCheckFlags uint32
+
+const (
+	checkAllowInteraction polkitCheckFlags = 0x01
+)
+
+type polkitAuthSubject struct {
+	Kind    string
+	Details map[string]dbus.Variant
+}
+
+type polkitAuthResult struct {
+	IsAuthorized bool
+	IsChallenge  bool
+	Details      map[string]string
+}
+
+// IsSenderAllowed returns nil if the user is allowed to perform a given operation.
+// Based on the D-Bus sender it will query the user's credentials and then
+// attempt to authorize the action using polkit.
+func (a Authorizer) IsSenderAllowed(ctx context.Context, action string, sender dbus.Sender) (err error) {
+	log.Debugf("Check if sender %s is allowed to perform action %q", sender, action)
+	defer decorate.OnError(&err, "permission denied")
+
+	credsResult := make(map[string]dbus.Variant)
+	if err = a.credsLookup.Call("org.freedesktop.DBus.GetConnectionCredentials", 0, string(sender)).Store(&credsResult); err != nil {
+		return err
+	}
+
+	var uid, pid uint32
+	uid, ok := credsResult["UnixUserID"].Value().(uint32)
+	if !ok {
+		return errors.New("can't get uid from dbus credentials")
+	}
+	pid, ok = credsResult["ProcessID"].Value().(uint32)
+	if !ok {
+		return errors.New("can't get pid from dbus credentials")
+	}
+
+	return a.isAllowed(ctx, action, pid, uid)
+}
+
+func (a Authorizer) isAllowed(ctx context.Context, action string, pid uint32, uid uint32) (err error) {
+	if uid == 0 {
+		log.Debug("Authorized as being administrator")
+		return nil
+	}
+
+	f, err := os.Open(filepath.Join(a.root, fmt.Sprintf("proc/%d/stat", pid)))
+	if err != nil {
+		return fmt.Errorf("couldn't open stat file for process: %w", err)
+	}
+	defer decorate.LogFuncOnErrorContext(ctx, f.Close)
+
+	startTime, err := getStartTimeFromReader(f)
+	if err != nil {
+		return err
+	}
+
+	subject := polkitAuthSubject{
+		Kind: "unix-process",
+		Details: map[string]dbus.Variant{
+			"pid":        dbus.MakeVariant(pid),
+			"start-time": dbus.MakeVariant(startTime),
+			"uid":        dbus.MakeVariant(uid),
+		},
+	}
+
+	var result polkitAuthResult
+	var details map[string]string
+	err = a.authority.Call(
+		"org.freedesktop.PolicyKit1.Authority.CheckAuthorization", dbus.FlagAllowInteractiveAuthorization,
+		subject, action, details, checkAllowInteraction, "").Store(&result)
+	if err != nil {
+		return fmt.Errorf("call to polkit failed: %w", err)
+	}
+	log.Debugf("Polkit call result, authorized: %t", result.IsAuthorized)
+
+	if !result.IsAuthorized {
+		return errors.New("polkit denied access")
+	}
+	return nil
+}
+
+// getStartTimeFromReader determines the start time from a process stat file content
+//
+// The implementation is intended to be compatible with polkit:
+//
+//	https://cgit.freedesktop.org/polkit/tree/src/polkit/polkitunixprocess.c
+func getStartTimeFromReader(r io.Reader) (t uint64, err error) {
```

Comment:
whoever wrote this was a complete genious :p

Reply:
roh, you don’t see the correct references :package: 

(TBH, this is a direct port of the C code, this is why I didn’t remember that was coming from snapd :p)